desmod/pool.py

```diff
@@ -0,0 +1,207 @@
+        self.pool = pool
+        pool._any_waiters.append(self)
+        pool._trigger_when_any()
+
+    def cancel(self):
+        if not self.triggered:
+            self.pool._any_waiters.remove(self)
+            self.callbacks = None
+
+
+class PoolWhenFullEvent(Event):
+    def __init__(self, pool):
+        super(PoolWhenFullEvent, self).__init__(pool.env)
+        self.pool = pool
+        pool._full_waiters.append(self)
+        pool._trigger_when_full()
+
+    def cancel(self):
+        if not self.triggered:
+            self.pool._full_waiters.remove(self)
+            self.callbacks = None
+
+
+class Pool(object):
+    """Simulation pool of arbitrary items.
+
+    `Pool` is similar to :class:`simpy.resources.Container`.
+    It provides a simulation-aware container for managing a pool of objects
+    needed by multiple processes.
+
+    Resources are added and removed using :meth:`put()` and :meth:`get()`.
+
+    :param env: Simulation environment.
+    :param capacity: Capacity of the pool; infinite by default.
+    :param hard_cap:
+        If specified, the pool overflows when the `capacity` is reached.
+    :param init_level: Initial level of the pool.
+    :param name: Optional name to associate with the queue.
+
+    """
+
+    def __init__(self, env, capacity=float('inf'), hard_cap=False,
+                 init_level=0, name=None):
+        self.env = env
+        #: Capacity of the queue (maximum number of items).
+        self.capacity = capacity
+        self._hard_cap = hard_cap
+        self.level = init_level
+        self.name = name
+        self._putters = []
+        self._getters = []
+        self._new_waiters = []
+        self._any_waiters = []
+        self._full_waiters = []
+        self._put_hook = None
+        self._get_hook = None
+        BoundClass.bind_early(self)
+
+    @property
+    def size(self):
+        """Number of items in pool."""
+        return self.level
+
+    @property
+    def remaining(self):
+        """Remaining pool capacity."""
+        return self.capacity - self.level
+
+    @property
+    def is_empty(self):
+        """Indicates whether the pool is empty."""
+        return self.level == 0
+
+    @property
+    def is_full(self):
+        """Indicates whether the pool is full."""
+        return self.level >= self.capacity
+
+    #: Put amount items in the pool.
+    put = BoundClass(PoolPutEvent)
+
+    #: Get amount items from the queue.
+    get = BoundClass(PoolGetEvent)
+
+    #: Return an event triggered when the pool is non-empty.
+    when_any = BoundClass(PoolWhenAnyEvent)
+
+    #: Return an event triggered when items are put in pool
+    when_new = BoundClass(PoolWhenNewEvent)
+
+    #: Return an event triggered when the pool becomes full.
+    when_full = BoundClass(PoolWhenFullEvent)
+
+    def _add_items(self, item_count=1):
+        self.level += item_count
+
+    def _remove_items(self, item_count=1):
+        self.level -= item_count
+        return item_count
```

Comment:
These `_add_items()` and `_remove_items()` methods only have one call site each and do a trivial operation. I'd suggest moving these `level` increments and decrements inline.